# %title%

**_Author: Carleton Smith_**

**Expected time = %expected_time%**

**Total points = 120 points**



    
## Assignment Overview

This assignment will both test your understand of linear algebra and your ability to translate procedures into code. Some questions will ask you to create your own implementations of linear algebra operations, while others will allow you to take advantage of existing implementations available in libraries, such as `numpy` and `scipy`.
In this assignment you will implement basic and intermediate linear algebra operations in Python.You will begin by performing matrix addition, substraction, and multplication. You will also use dot product and cross product to perform calculations. You will then move on to more advanced operations such as transpotion, matrix determinant, and inverse. 

This assignment is designed to build your familiarity and comfort coding in Python while also helping you review key topics from each module. As you progress through the assignment, answers will get increasingly complex. It is important that you adopt a data scientist's mindset when completing this assignment. **Remember to run your code from each cell before submitting your assignment.** Running your code beforehand will notify you of errors and give you a chance to fix your errors before submitting. You should view your Vocareum submission as if you are delivering a final project to your manager or client. 

***Vocareum Tips***
- Do not add arguments or options to functions unless you are specifically asked to. This will cause an error in Vocareum.
- Do not use a library unless you are expicitly asked to in the question. 
- You can download the Grading Report after submitting the assignment. This will include feedback and hints on incorrect questions. 

### Learning Objectives

- Apply foundational linear algebra concepts in python to build basic algorithims.
- Use matrix operations to solve equations. 
- Perform advanced linear algebra procedures, such as PCA and Singular Value Decompositon (SVD)
- Apply linear algebra to solve least squares linear regression




## Index:

#### %title%

##### %subtitle_1%
- [Question 1](#Question-1)
- [Question 2](#Question-2)
- [Question 3](#Question-3)
- [Question 4](#Question-4)
- [Question 5](#Question-5)

##### %subtitle_2%
- [Question 6](#Question-6)
- [Question 7](#Question-7)
- [Question 8](#Question-8)
- [Question 9](#Question-9)

##### %subtitle_3%
- [Question 10](#Question-10)
- [Question 11](#Question-11)

##### %subtitle_4%
- [Question 12](#Question-12)



## %title%

**Implement fundamental linear algebra procedures with Python**

This assignment will test your ability to demonstrate your understanding of fundamental linear algebra procedures and your ability to translate these procedures into code. Many (if not all) of the linear algebra procedures discussed in this assignment have existing implementations in Python. In practice, you will leverage these prebuilt implementations, but for the purpose of understanding the concepts and practicing Python, you will be asked to produce your own implementations.

The below functions are helper functions to assist with various tasks throughout the assignment. You can ignore them.

In [1]:
import numpy as np
import random
import inspect
import statsmodels.api as sm
from scipy.linalg import svd
from sklearn.datasets import make_regression
import matplotlib.pyplot as plt

In [2]:
def test_for_banned_funcs_and_oper(func, banned_funcs, banned_oper=[]):
    '''This function will test if your code uses a forbidden function'''
    unbound_closures = inspect.getclosurevars(func).unbound
    
    # check for banned operators
    student_code = inspect.getsourcelines(func)[0]
    if banned_oper:
        for line in student_code:
            for op in banned_oper:
                assert op not in line, "Cannot use operator(s) like '{}' in solution.".format(op)
    
    funcs_in_violation = []
    for uc in unbound_closures:
        if uc in banned_funcs:
            if uc == 'T':
                uc = '.T'
            funcs_in_violation.append(uc)
    assert len(funcs_in_violation) == 0,"Cannot use function(s) '{}' in solution.".format("' or '".join(funcs_in_violation))


In [3]:
def matrix_maker(*args, disp=True):
    '''This function will create an arbitrary number of matrices of specific shapes'''
    all_mats = []
    if not len(args) % 2:
        num_mats = int(len(args)/2)
        for _ in np.arange(num_mats): 
            n, m, *args = args
            if not m:
                M = np.random.randint(10, size=(n,))
            else:
                M = np.random.randint(10, size=(n,m))
            all_mats.append(M)
        if disp:
            for mat in all_mats:
                print(mat)
        if len(all_mats) == 1:
            return all_mats[0]
        return tuple(all_mats)
    else:
        raise ValueError("Must give even number of dimensions for each matrix.")



## %subtitle_1%
The following questions will ask you to create simple functions that perform basic matrix operations.

[Back to top](#Index:) 

### Question 1

*4 Points*

Your first task is to create a function called `matrix_adder` that will accept two matrices of the same shape as parameters and return another numpy array that is the result of matrix addition.

    Parameters:
    A -- numpy array of shape (n,m)
    B -- numpy array of shape (n,m)
    
    Returns: a numpy array of shape (n,m) that is the sum of A + B.
    
    *If input matrices are of invalid shape, return 'Invalid dimensions'.*

Use the matrices `A` and `B` below as test cases.

In [4]:
A = np.array([
    [7, 2, 0, 0],
    [8, 8, 1, 9],
    [8, 6, 6, 8]
])

B = np.array([
    [7, 7, 8, 5],
    [5, 2, 5, 0],
    [0, 5, 5, 2]
])

In [5]:
### GRADED

### YOUR SOLUTION HERE
def matrix_adder(matrix1, matrix2):
    """Perform matrix addition with two input matrices.
    
    Parameters:
    matrix1 -- numpy array of shape (n,m)
    matrix2 -- numpy array of shape (n,m)
    
    Returns: a numpy array of shape (n,m); the result of matrix addition.
    
    If input matrices are of invalid shape, return 'Invalid dimensions'.
    """
    return

### BEGIN SOLUTION
def matrix_adder(matrix1, matrix2):
    if matrix1.shape != matrix2.shape:
        return "Invalid dimensions"
    return matrix1 + matrix2
### END SOLUTION

### For verifying answer:
print('-'*25)
print("Input Matrices:")
print(A)
print(B)
print('-'*25)
print("Your Solution:")
print(matrix_adder(A, B))

# wrong dimensions
A2, B2 = matrix_maker(3,5,3,4, disp=False)
print('-'*25)
print("Invalid dimension case:\t{}".format(matrix_adder(A2, B2)))

-------------------------
Input Matrices:
[[7 2 0 0]
 [8 8 1 9]
 [8 6 6 8]]
[[7 7 8 5]
 [5 2 5 0]
 [0 5 5 2]]
-------------------------
Your Solution:
[[14  9  8  5]
 [13 10  6  9]
 [ 8 11 11 10]]
-------------------------
Invalid dimension case:	Invalid dimensions


In [6]:
### BEGIN HIDDEN TESTS (4)
import random
import numpy as np

def matrix_adder_(matrix1_, matrix2_):
    if matrix1_.shape != matrix2_.shape:
        return "Invalid dimensions"
    return matrix1_ + matrix2_



for i in range(5):
    n = random.randint(1,i+4)
    m = random.randint(1,i+4)
    A_, B_ = matrix_maker(n,m,n,m, disp=False)
    ans_ = matrix_adder_(A_, B_)
    ans = matrix_adder(A_, B_)
    res_type = type(matrix_adder_(A_, B_))
    assert isinstance(matrix_adder(A_, B_), type(matrix_adder_(A_, B_))), f"Ensure you return an array of type '{res_type}'"
    np.testing.assert_almost_equal(ans, ans_)
    
# test for behavior when sizes don't match
n = random.randint(1,8)
m = random.randint(1,8)
A_, B_ = matrix_maker(n,m-1,n,m, disp=False)
ans_ = matrix_adder_(A_, B_)
ans = matrix_adder(A_, B_)

assert 'invalid' in ans.strip().lower(), "Make sure to return the string 'Invalid dimensions' when arrays have different shapes. Check spelling."
print("That's correct")

### END HIDDEN TESTS

That's correct


[Back to top](#Index:) 

### Question 2

*4 Points*

Similarly to Q1, create a function called `matrix_subtraction` that will accept two matrices of the same shape as parameters and return another numpy array that is the result of matrix subtraction.

    Parameters:
    A -- numpy array of shape (n,m)
    B -- numpy array of shape (n,m)
    
    Returns: a numpy array of shape (n,m) that is the difference of A and B.

Use the matrices `A` and `B` below as test cases.

In [7]:
A = np.array([
    [0, 8, 6, 4],
    [2, 9, 0, 9],
    [1, 2, 1, 9]
])

B = np.array([
    [3, 4, 4, 3],
    [2, 9, 2, 9],
    [0, 2, 6, 5]
])

In [8]:
### GRADED
def matrix_subtraction(matrix1, matrix2):
    """Perform matrix subtraction with two input matrices.
    
    Parameters:
    matrix1 -- numpy array of shape (n,m)
    matrix2 -- numpy array of shape (n,m)
    
    Returns: a numpy array of shape (n,m); the result of matrix subtraction.
    
    If input matrices are of invalid shape, return 'Invalid dimensions'.
    """
    return

### BEGIN SOLUTION
def matrix_subtraction(matrix1, matrix2):
    if matrix1.shape != matrix2.shape:
        return "Invalid dimensions"
    return matrix1 - matrix2
### END SOLUTION

### For verifying answer:
print('-'*25)
print("Input Matrices:")
print(A, B)
print('-'*25)
print("Your Solution:")
print(matrix_subtraction(A, B))

# wrong dimensions
A2, B2 = matrix_maker(3,5,3,4, disp=False)
print('-'*25)
print("Invalid dimension case:\t{}".format(matrix_subtraction(A2, B2)))

-------------------------
Input Matrices:
[[0 8 6 4]
 [2 9 0 9]
 [1 2 1 9]] [[3 4 4 3]
 [2 9 2 9]
 [0 2 6 5]]
-------------------------
Your Solution:
[[-3  4  2  1]
 [ 0  0 -2  0]
 [ 1  0 -5  4]]
-------------------------
Invalid dimension case:	Invalid dimensions


In [9]:
### BEGIN HIDDEN TESTS (4)
import random
import numpy as np

def matrix_subtraction_(matrix1_, matrix2_):
    if matrix1_.shape != matrix2_.shape:
        return "Invalid dimensions"
    return matrix1_ - matrix2_

for i in range(5):
    n = random.randint(1,i+4)
    m = random.randint(1,i+4)
    A_, B_ = matrix_maker(n,m,n,m,disp=False)
    ans_ = matrix_subtraction_(A_, B_)
    ans = matrix_subtraction(A_, B_)
    res_type = type(matrix_subtraction_(A_, B_))
    assert isinstance(matrix_subtraction(A_, B_), type(matrix_subtraction_(A_, B_))), f"Ensure you return an array of type '{res_type}'"
    np.testing.assert_almost_equal(ans, ans_)
    
# test for behavior when sizes don't match
n = random.randint(1,8)
m = random.randint(1,8)
A_, B_ = matrix_maker(n,m-1,n,m, disp=False)
ans_ = matrix_subtraction_(A_, B_)
ans = matrix_subtraction(A_, B_)

assert 'invalid' in ans.strip().lower(), "Make sure to return the string 'Invalid dimensions' when arrays have different shapes. Check spelling."
print("That's correct!")

### END HIDDEN TESTS

That's correct!


[Back to top](#Index:) 

### Question 3

*8 Points*

Your next task will be to implement another routine task in linear algebra - transpose a matrix. **For this exercise, you are not allowed to use any existing function or method to produce the transpose**. The reason for this limitation is to provide opportunities to practice manipulating `numpy` arrays natively.

Examples of fobidden functions:
    - numpy.transpose
    - numpy.matrix.transpose
    - numpy.ndarray.T
    - pandas.DataFrame.T
    - pandas.DataFrame.transpose

Create a function called `transpose` that will accept a single matrix (all numeric - int or float) of shape (n,m) as a parameter and return the transpose of that matrix as a numpy array of shape (m,n).

    Parameters:
    A -- numpy array of shape (n,m)
    
    Returns: a numpy array of shape (m,n) that is the transpose of A.

Use the matrices `A` below as a test case.

In [10]:
A = np.array([
    [6, 6, 0, 8],
    [7, 8, 3, 2],
    [7, 4, 8, 6]
])

In [11]:
### GRADED
def transpose(X):
    """Perform matrix transpose.
    
    Parameters:
    X -- numpy array of shape (n,m)
    
    Returns: a numpy array of shape (m,n); the transpose of matrix X.
    
    If X is 1 dimensional, return the original array.
    """
    return

### BEGIN SOLUTION
def transpose(X):
    # check if X is one dimensional
    if len(X.shape) == 1:
        return X
    
    n_rows = X.shape[0] # number of rows
    n_cols = X.shape[1] # number of columns
    
    transpose_arr = np.empty(shape=(n_cols, n_rows)) # initialize empty array
    
    # iterate through the rows and make them columns
    for row_idx, row in enumerate(X):
        transpose_arr[:, row_idx] = X[row_idx, :]
    return transpose_arr
    
    
### END SOLUTION

### For verifying answer:
print('-'*25)
print("Input Matrix:")
print(A)
print('-'*25)
print("Your Solution:")
print(transpose(A))

# one dimension array validation
A2 = matrix_maker(3,None,disp=False)
print('-'*25)
print("Input array:\t\t{}".format(A2))
print("One dimesional case:\t{}".format(transpose(A2)))

-------------------------
Input Matrix:
[[6 6 0 8]
 [7 8 3 2]
 [7 4 8 6]]
-------------------------
Your Solution:
[[6. 7. 7.]
 [6. 8. 4.]
 [0. 3. 8.]
 [8. 2. 6.]]
-------------------------
Input array:		[5 8 8]
One dimesional case:	[5 8 8]


In [12]:
### BEGIN HIDDEN TESTS (8)
import random
import inspect
import numpy as np

def transpose_(X_):
    if len(X_.shape) == 1:
        return X_
    n_rows_ = X_.shape[0]
    n_cols_ = X_.shape[1]
    transpose_arr_ = np.empty(shape=(n_cols_, n_rows_))
    
    # iterate through the rows and make them columns
    for row_idx, row in enumerate(X_):
        transpose_arr_[:, row_idx] = X_[row_idx, :]
    return transpose_arr_

# test for "illegal" use of numpy/pandas functions and operators
banned = ['.transpose(', '.T(', '.T']
banned_funcs = ['transpose', 'T']
banned_oper = []
#test_for_banned_funcs_and_oper(transpose, banned_funcs, banned_oper)
#for idx, line in enumerate(inspect.getsourcelines(transpose)[0]):
#for f in banned:
 #       if f in ('.T'):
  #          assert f not in line, "Cannot use function '{}' in line {} of solution.".format(f, idx)
   #     else:
    #        assert f not in line, "Cannot use function '{}' in line {} of solution.".format(f+')', idx)

for i in range(5):
    n = random.randint(1,i+4)
    m = random.randint(1,i+4)
    A_= matrix_maker(n,m,disp=False)
    ans_ = transpose_(A_)
    ans = transpose(A_)
    res_type = type(transpose_(A_))
    assert isinstance(transpose(A_), res_type), f"Ensure you return an array of type '{res_type}'"
    np.testing.assert_almost_equal(ans, ans_)
    
# test for behavior when matrix 1D
n = random.randint(1,8)
m = random.randint(1,8)
A_ = matrix_maker(n,None,disp=False)
ans_ = transpose_(A_)
ans = transpose(A_)

np.testing.assert_equal(ans, ans_, err_msg = "The matrix is not correct", verbose = False)
print("That's correct!")


### END HIDDEN TESTS

That's correct!


[Back to top](#Index:) 

### Question 4

*9 Points*

Create a function called `dot_prod` that computes the dot product between two 1 dimensional vectors and return the result as an integer.

To compute the dot product, consider the two vectors below:

$$a = \begin{bmatrix} a_1 \\ \vdots \\ a_n \end{bmatrix}     b = \begin{bmatrix} b_1 \\ \vdots \\ b_n \end{bmatrix}$$

The dot product calculation is as follows:

$$a \cdot b = a_1b_1 + ... + a_nb_n$$

**For this exercise, you are not allowed to use any existing function or method to produce the dot product**. For the same reason mentioned in the earlier, this limitation is in place to enable coding practice, as well as reinforcing the mechanics of the linear algebra calculation.

Examples of fobidden functions:
    - numpy.dot()
    - numpy.inner()
    - numpy.matmul()
    - pandas.DataFrame.dot()
    - pandas.Series.dot()
    - the `@` operator

    Parameters:
    vecA -- a 1 dimensional numpy array of shape (n,)
    vecB -- a 1 dimensional numpy array of shape (n,)
    
    Returns: int; the dot product of the vectors vecA and vecB

Use the matrices `A` below as a test case.

In [13]:
a = np.array([3, 4, 2])
b = np.array([5, 5, 0])

In [14]:
### GRADED

### YOUR SOLUTION HERE
def dot_prod(vecA, vecB):
    """Calculate the dot product between two 1D vectors.
    
    Parameters:
    vecA -- a 1 dimensional numpy array of shape (n,)
    vecB -- a 1 dimensional numpy array of shape (n,)
    
    Returns: int or float; the dot product of the vectors vecA and vecB
    """
    return

### BEGIN SOLUTION
def dot_prod(vecA, vecB):
    return sum(vecA*vecB)


### END SOLUTION

### For verifying answer:
print('-'*25)
print("Vector A:")
print(a)
print('-'*25)
print("Vector B:")
print(b)
print('-'*25)
print("Your Solution: {}".format(dot_prod(a,b)))
print("True Solution: {}".format(np.dot(a,b)))

-------------------------
Vector A:
[3 4 2]
-------------------------
Vector B:
[5 5 0]
-------------------------
Your Solution: 35
True Solution: 35


In [15]:
### BEGIN HIDDEN TESTS (9)
import random
import inspect
import numpy as np

def dot_prod_(vecA_, vecB_):
    return sum(vecA_*vecB_)


# test for "illegal" use of numpy/pandas functions and operators
banned = ['.dot(', '.inner(', '.matmul(',]
banned_funcs = ['dot', 'inner', 'matmul',]
banned_oper = ['@',]
#test_for_banned_funcs_and_oper(dot_prod, banned_funcs, banned_oper)
#for idx, line in enumerate(inspect.getsourcelines(dot_prod)[0]):
#for f in banned:
 #       if f in ('.T'):
  #          assert f not in line, "Cannot use function '{}' in line {} of solution.".format(f, idx)
   #     else:
    #        assert f not in line, "Cannot use function '{}' in line {} of solution.".format(f+')', idx)

# test solution
for i in range(5):
    n = random.randint(1,i+4)
    A_, B_ = matrix_maker(n, None, n, None, disp=False)
    ans_ = dot_prod_(A_, B_)
    ans = dot_prod(A_, B_)
    np.testing.assert_equal(ans, ans_), "The result of your dot poduct is not correct"
print("That's correct")
### END HIDDEN TESTS

That's correct


[Back to top](#Index:) 

### Question 5

*10 Points*

Create a function that performs matrix multiplication between two 2 dimensional matrices. Your function should accept two matrices of appropriate dimensions and return a numpy array that is the result of matrix multiplication. See [this resource](https://www.mathsisfun.com/algebra/matrix-multiplying.html) for a simple tutorial.

    Parameters:
    A -- numpy array of shape (n,m)
    B -- numpy array of shape (m,p)
    
    Returns: a numpy array of shape (n,p) that is the solution of AB matrix multiplication

**Your function should:**

1. Require `A` and `B` to have the acceptable dimensions for matrix multiplication.

    - If A or B have incompatible dimensions, your function should **return** the string `"Incorrect Dimensions"`

    
2. Avoid using the following functions:


    - numpy.dot()
    - numpy.inner()
    - numpy.matmul()
    - pandas.DataFrame.dot()
    - pandas.Series.dot()
    - the `@` operator


Use the example matrices below for testing.

In [16]:
A = np.array([
    [2,1],
    [1,2],
    [3,2]
])

B = np.array([
    [-1, 4],
    [1, -3]
])

In [17]:
### GRADED

### YOUR SOLUTION HERE
def matrix_mult(A, B):
    """Perform matrix multiplication with matrices A and B.
    
    Parameters:
    A -- numpy array of shape (n,m)
    B -- numpy array of shape (m,p)
    
    Returns:
    a numpy array of shape (n,p); the result of AB matrix multiplication
    """
    return

### BEGIN SOLUTION
def matrix_mult(A,B):
    n, a_m = A.shape  # num rows/cols in A
    b_m, p = B.shape  # num rows/cols in B
    
    # check if dimesions are right for matrix multiplication
    if a_m != b_m:
        return "Incorrect Dimensions"
    
    result = np.empty(shape=(n, p))  # empty array for result
    row_ids = range(n)  # row indices in result
    col_ids = range(p)  # col indices in result
    
    # iterate through rows and columns & perform calculation
    for row in row_ids:
        for col in col_ids:
            result[row, col] = sum(A[row, :] * B[:, col])
    return result
### END SOLUTION

In [18]:
### BEGIN HIDDEN TESTS (10)
def matrix_mult_(A_,B_):
    n, a_m = A_.shape  # num rows/cols in A
    b_m, p = B_.shape  # num rows/cols in B
    
    # check if dimesions are right for matrix multiplication
    if a_m != b_m:
        return "Incorrect Dimensions"
    
    result = np.empty(shape=(n, p))  # empty array for result
    row_ids = range(n)  # row indices in result
    col_ids = range(p)  # col indices in result
    
    # iterate through rows and columns & perform calculation
    for row in row_ids:
        for col in col_ids:
            result[row, col] = sum(A_[row, :] * B_[:, col])
    return result

# test for "illegal" use of numpy/pandas functions and operators
banned_funcs = ['matmul', 'dot', 'T']
banned_oper = ['@']
#test_for_banned_funcs_and_oper(matrix_mult, banned_funcs, banned_oper)

# test the output is correct
n = np.random.randint(2,10)
m = np.random.randint(2,10)
p = np.random.randint(2,10)
A_ = np.random.randint(10, size=(n, m))
B_ = np.random.randint(10, size=(m, p))
solution = matrix_mult_(A_, B_)
student = matrix_mult(A_, B_)

np.testing.assert_array_equal(solution, student, err_msg = "Your function does not return the correct results", verbose=False)
print("That's correct")
### END HIDDEN TESTS

That's correct


[Back to top](#Index:) 

## %subtitle_2%

### Question 6

*10 Points*

Create a function called `determinant2x2` that will calculate the determinant of any 2x2 matrix. You may use the 2x2 matrix, `A2`, below as a test case. See [this resource](https://www.mathsisfun.com/algebra/matrix-determinant.html) for a simple tutorial.

    
    Parameters:
    M -- numpy array of shape (n,n)
    
    Returns: float; the determinant of M.
    
For this exercise, avoid using the `numpy.linalg.det()` numpy function.

In [19]:
A2 = np.array([
    [5.5, 7],
    [2, 1]
])

In [20]:
### GRADED

### YOUR SOLUTION HERE
def determinant2x2(M):
    """Calculate the determinant of the 2x2 matrix, M.
    
    Parameters:
    M -- numpy array of shape (n,n)
    
    Returns: float; the determinant of M.
    """
    return

### BEGIN SOLUTION
def determinant2x2(M):
    det = (M[0,0]*M[1,1] - M[0,1]*M[1,0])
    return float(det)
### END SOLUTION

### For verifying answer:
print('-'*20)
print("Sample Matrix:")
A2 = matrix_maker(2,2)
print('-'*20)
print("Numpy Solution:\t{}".format(np.linalg.det(A2)))
print("Your Solution:\t{}".format(determinant2x2(A2)))
print()
print('(Rounding errors are not an issue with the grader)')

--------------------
Sample Matrix:
[[6 7]
 [2 4]]
--------------------
Numpy Solution:	10.000000000000002
Your Solution:	10.0

(Rounding errors are not an issue with the grader)


In [21]:
### BEGIN HIDDEN TESTS (10)
def determinant2x2_(M_):
    det_ = (M_[0,0]*M_[1,1] - M_[0,1]*M_[1,0])
    return float(det_)

# test for "illegal" use of numpy/pandas functions and operators
banned_funcs = ['det']
banned_oper = []
#test_for_banned_funcs_and_oper(determinant2x2, banned_funcs, banned_oper)

# test the output is correct

A_ = np.random.randint(10, size=(2, 2))
solution = determinant2x2_(A_)
student = determinant2x2(A_)

np.testing.assert_equal(solution, student,  err_msg = "The result retirn by your function is not correct", verbose=False)
print("That's correct")

### END HIDDEN TESTS

That's correct


[Back to top](#Index:) 

### Question 7

*15 Points*

With the 3x3 matrix, `A2`, create a function called `determinant3x3` that will calculate the determinant of any 3x3 matrix. See [this resource](https://www.mathsisfun.com/algebra/matrix-determinant.html) for a simple tutorial.

For this exercise, avoid using the `numpy.linalg.det()` numpy function. You may, however, use the functions created from Question 6 as part of your solution.

In [22]:
A3 = np.array([
    [7, 8, 1],
    [5, 5, 5],
    [7, 0, 3]
])

In [23]:
### GRADED

### YOUR SOLUTION HERE
def determinant3x3(M):
    return

### BEGIN SOLUTION
def determinant3x3(M):
    bot_right_det = determinant2x2(M[1:,1:])
    sides_arr = M[np.array([1,1,2,2]),np.array([0,2,0,2])].reshape(2,2) # fancy indexing, see numpy docs
    sides_det = determinant2x2(sides_arr)
    bot_left_det = determinant2x2(M[1:,0:2])
    dets = [bot_right_det, sides_det, bot_left_det]
    for idx, val, det in zip(np.arange(3), M[0], dets):
        if idx == 0:
            deter = val*det
        elif idx == 1:
            deter -= val*det
        else:
            deter += val*det
    return float(deter)
### END SOLUTION

### For verifying answer:
print('-'*20)
print("Sample Matrix:")
A3 = matrix_maker(3,3)
print('-'*20)
print("Numpy Solution:\t{}".format(np.linalg.det(A3)))
print("Your Solution:\t{}".format(determinant3x3(A3)))
print()
print('(Rounding errors are not an issue with the grader)')

--------------------
Sample Matrix:
[[6 4 3]
 [4 5 9]
 [4 5 4]]
--------------------
Numpy Solution:	-70.00000000000003
Your Solution:	-70.0

(Rounding errors are not an issue with the grader)


In [24]:
### BEGIN HIDDEN TESTS (15)
def determinant2x2_(M_):
    det_ = (M_[0,0]*M_[1,1] - M_[0,1]*M_[1,0])
    return float(det_)


def determinant3x3_(M_):
    bot_right_det_ = determinant2x2_(M_[1:,1:])
    sides_arr_ = M_[np.array([1,1,2,2]),np.array([0,2,0,2])].reshape(2,2)
    sides_det_ = determinant2x2_(sides_arr_)
    bot_left_det_ = determinant2x2_(M_[1:,0:2])
    dets_ = [bot_right_det_, sides_det_, bot_left_det_]
    for idx_, val_, det_ in zip(np.arange(3), M_[0], dets_):
        if idx_ == 0:
            deter_ = val_*det_
        elif idx_ == 1:
            deter_ -= val_*det_
        else:
            deter_ += val_*det_
    return float(deter_)

# test for "illegal" use of numpy/pandas functions and operators
banned_funcs = ['det']
banned_oper = []
#test_for_banned_funcs_and_oper(determinant3x3, banned_funcs, banned_oper)

# test the output is correct
A_ = np.random.randint(10, size=(3, 3))
solution = determinant3x3_(A_)
student = determinant3x3(A_)

np.testing.assert_equal(solution, student, err_msg = "The determinant is not correct", verbose=False)
print("That's correct!")

### END HIDDEN TESTS

That's correct!


[Back to top](#Index:) 

## Linear Independence

### Question 8

*10 Points*

For this question, refer to the lecture section on Linear Independence and the supplemental reading. Recall the definitions of dependence and independence:

**Dependence** means that at least one equation can be algebraically derived from the others (i.e., can be written as a linear combination of other equations). There are an **infinite** number of solutions that will satisfy the conditions of the equations! To know which solution you want, you have to feed in an x value. This makes the y value dependent on the x value. Attempting to solve a dependent system of linear equations via `numpy` will throw an error, because the solution depends on your inputs.

**Independence** means that no equation in the system can be algebraically derived from the others. In systems of linear equations with 2 equations, this means that the lines only intersect at one point, or the equations reflect parallel lines. There is either **no solution** or a **single solution**.

**Supplemental Reading**
* [Systems of Linear Equations](https://www.impan.pl/~pmh/teach/algebra/additional/lineareq.pdf)

<img src="assets/linear_independence.png" width="500">

Consider the following sets of systems of equations

**System 1**

$$A1 = \begin{bmatrix}7 & 0 & 0 \\ 4 & 6 & 3 \\ 0 & 4 & 2\\ \end{bmatrix} ;   b1 = \begin{bmatrix}7 \\ 0 \\ 0 \end{bmatrix}$$


**System 2**

$$A2 = \begin{bmatrix}7 & 2 & 9 \\ 8 & 5 & 1 \\ 1 & 2 & 8\\ \end{bmatrix};    b2 = \begin{bmatrix}6 \\ 8 \\ 4 \end{bmatrix}$$


Your task for this question is as follows:
 - Determine which system of equations above is linearly independent. Indicate which system is linearly independent by assigning the string `"system 1"` or `"system 2"` to the variable `ans8a`.
 - For the system that is linearly independent, solve the system of equations and assign the solution to the variable `ans8b`. Round all values to 3 decimal places.
 
Note: There are no restrictions on using functions from Python libraries for this question.

In [25]:
# system of equations 1
A1 = np.array([
    [7, 0, 0],
    [4, 6, 3],
    [0, 4, 2]])
b1 = np.array([4, 8, 5])

# system of equations 2
A2 = np.array([
    [7, 2, 9],
    [8, 5, 1],
    [1, 2, 8]])
b2 = np.array([6, 8, 4])

In [26]:
### GRADED

### YOUR SOLUTION HERE
ans8a = None
ans8b = None


### BEGIN SOLUTION
# determine which system is indep.
# syst_1 = np.linalg.solve(A1, b1) <-- throws error
syst_2 = np.linalg.solve(A2, b2)
ans8a = "system 2"
ans8b = syst_2
### END SOLUTION

### For verifying answer:
print("Linear Independent System:\t'{}'".format(ans8a))
print("Solution for '{}':\t{}".format(ans8a, ans8b))

Linear Independent System:	'system 2'
Solution for 'system 2':	[0.30125523 1.07949791 0.19246862]


In [27]:
### BEGIN HIDDEN TESTS (10)
# system of equations 2
A2_ = np.array([
    [7, 2, 9],
    [8, 5, 1],
    [1, 2, 8]])
b2_ = np.array([6, 8, 4])

syst_2_ = np.linalg.solve(A2_, b2_)
ans8a_ = "system2"
ans8b_ = syst_2_

assert type(ans8a_) == type(ans8a), "Ensure you provide `ans8a` as a string. Answer choices are 'system 1' or 'system 2'."
assert ans8a_ == ans8a.strip().lower().replace(' ',''), "Ensure your spelling for `ans8a` is correct."
syst_2_ = np.linalg.solve(A2_, b2_)
ans8b_ = syst_2_
np.testing.assert_equal(ans8b_, ans8b, err_msg = "The array are not the same", verbose = False)
print("That's correct!")
### END HIDDEN TESTS

That's correct!


[Back to top](#Index:) 

## %subtitle_3%
Eigendecomposition is a well defined linear algebra procedure that is useful for dimensionality reduction.

<img src="assets/eigenvalue.svg" width="400">

For example, Principal Component Analysis (PCA) leverages eigendecomposition to reduce dimensionality by extracting new features (called components) that maximize the variance in the data.

The process for performing PCA is as follows:
1. Center the data around the origin by subtracting each value from its column mean
2. Compute the covariance matrix (square matrix)
3. Perform eigendecomposition on covariance matrix found in step 2.
4. Project data into subspace by taking the dot product of the transposed eigenvectors with the transposed mean centered data from step 1.

This process is demonstrated in the PCA lesson material.

[Back to top](#Index:) 

### Question 9

*15 Points*

For the following question, you will use the matrix `A` defined in the cell below. Do not round any values or answers.

Your task is as follows:
- Center the data around the origin by subtracting each value from it's column mean (computed below and bound to `col_means`). Assign this matrix to `ans9a`.
- Compute the covariance matrix (square matrix) from step 1. Assign this result to the variable `ans9b`.
- Perform eigendecomposition on covariance matrix found in step 2. Assign the vectors and values to the vaariables `ans9c` and `ans9d`, respectively.
- Project data into subspace by taking the dot product of the transposed eigenvectors with the transposed mean centered data from step 1. Assign the resulting matrix to the variable `ans9e`.

In [28]:
A, _ = make_regression(n_samples=1000, n_features = 5, noise=5, random_state=24)
print(A[:5])

[[ 0.33971681 -0.7452978   0.71999357 -0.10898928 -0.80510445]
 [-1.59009262  0.34321702 -1.89241933  1.7576691   1.15028127]
 [ 0.26459075 -1.24608616  0.74371496 -0.66519186 -0.97569435]
 [ 1.14975178  0.92126388  0.88806273 -0.24656481  0.49593947]
 [-0.96989669  0.85091215 -0.56442506 -0.6248215  -0.88181707]]


In [29]:
### GRADED

### YOUR SOLUTION HERE
col_means = np.round(np.mean(A, axis=0), 3) # mean of each column
ans9a = None
ans9b = None
ans9c , ans9d = None , None # hint: np.linalg.eig()
ans9e = None

### BEGIN SOLUTION
# compute the mean for each column
col_means = np.mean(A, axis=0)
# center data around column mean
ans9a = A - col_means
# compute covariance matrix
ans9b = np.cov(ans9a)
# compute eigenvalues, eigenvectors from cov matrix
ans9c, ans9d = np.linalg.eig(ans9b)
# project into subspace
ans9e = ans9d.T.dot(ans9a)
### END SOLUTION

In [30]:
### BEGIN HIDDEN TESTS (15)
A_, _ = make_regression(n_samples=1000, n_features = 5, noise=5, random_state=24)
col_means_ = np.mean(A_, axis=0)
ans9a_ = A_ - col_means_
ans9b_ = np.cov(ans9a_)
ans9c_, ans9d_ = np.linalg.eig(ans9b_)
ans9e_ = ans9d_.T.dot(ans9a_)

student = [ans9a, ans9b, ans9c, ans9d, ans9e]
solution = [ans9a_, ans9b_, ans9c_, ans9d_, ans9e_]

for stu, sol in zip(student, solution):
    np.testing.assert_equal(stu, sol, err_msg = "Your decomposition is not correct", verbose = False)
print("That's correct!")
### END HIDDEN TESTS

That's correct!


[Back to top](#Index:) 

### Question 10

*10 Points*

Now that you have implemented PCA from scratch, this question tasks you with implementing PCA using `scikit-learn`.

- Instantiate an instance of the class `PCA` with the default parameters. Bind this object to the variable `ans10a`.
- With the instantiated PCA object, fit and transform the matrix `A` (defined below). Bind the transformed data to the variable `ans10b`.
- What is the cumulative ratio of explained variance of the **first two components**? Assign the cumulative ratio of explained variance to the variable `ans10c` as a float. Round to 2 decimal places.

Hint: the `.explained_variance_ratio_` attribute may be helpful. See [docs](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html#sklearn-decomposition-pca) for more.

In [31]:
from sklearn.decomposition import PCA
A, _ = make_regression(n_samples=1000, n_features = 5, noise=5, random_state=24)
print(A[:5])

[[ 0.33971681 -0.7452978   0.71999357 -0.10898928 -0.80510445]
 [-1.59009262  0.34321702 -1.89241933  1.7576691   1.15028127]
 [ 0.26459075 -1.24608616  0.74371496 -0.66519186 -0.97569435]
 [ 1.14975178  0.92126388  0.88806273 -0.24656481  0.49593947]
 [-0.96989669  0.85091215 -0.56442506 -0.6248215  -0.88181707]]


In [32]:
### GRADED

### YOUR SOLUTION HERE
ans10a = None
ans10b = None
ans10c = None

### BEGIN SOLUTION
ans10a = PCA()
ans10b = ans10a.fit_transform(A)
ans10c = np.round(np.cumsum(ans10a.explained_variance_ratio_)[1], 2)
### END SOLUTION

### For verifying answer:
print("PCA Instance:")
print()
print(ans10a)
print('-'*45)
print("Original Matrix:")
print(A[:5])
print('-'*45)
print("PCA Transformed Matrix:")
print(ans10b[:5])
print('-'*45)
print("Ratio of Explained Variance for Components 1 & 2: {}".format(ans10c))

PCA Instance:

PCA()
---------------------------------------------
Original Matrix:
[[ 0.33971681 -0.7452978   0.71999357 -0.10898928 -0.80510445]
 [-1.59009262  0.34321702 -1.89241933  1.7576691   1.15028127]
 [ 0.26459075 -1.24608616  0.74371496 -0.66519186 -0.97569435]
 [ 1.14975178  0.92126388  0.88806273 -0.24656481  0.49593947]
 [-0.96989669  0.85091215 -0.56442506 -0.6248215  -0.88181707]]
---------------------------------------------
PCA Transformed Matrix:
[[ 0.64076549  0.75184713  0.54150641 -0.66347072  0.35965772]
 [-1.89895284 -1.22622239  0.85221523  1.88415338 -1.0866549 ]
 [ 0.52870039  0.94079611  0.59117183 -0.97437884  1.0290358 ]
 [ 0.78541905  0.1556773  -1.29146843 -0.61474522 -0.83824043]
 [ 0.542303   -0.53253399 -0.1541736   1.24274883  0.99817106]]
---------------------------------------------
Ratio of Explained Variance for Components 1 & 2: 0.44


In [33]:
### BEGIN HIDDEN TESTS (10)
A_, _ = make_regression(n_samples=1000, n_features = 5, noise=5, random_state=24)
ans10a_ = PCA()
ans10b_ = ans10a_.fit_transform(A_)
ans10c_ = np.round(np.cumsum(ans10a_.explained_variance_ratio_)[1], 2)

assert ans10a_.get_params() == ans10a.get_params(), "Instantiate PCA() with default parameters."
np.testing.assert_equal(ans10b, ans10b_, err_msg = "The matrix is not correct", verbose = False)
np.testing.assert_equal(ans10c, ans10c_, err_msg = "The matrix is not correct", verbose = False)
print("That's correct!")

### END HIDDEN TESTS

That's correct!


[Back to top](#Index:) 

### Question 11

*10 Points*

Another linear algebra technique for reducing dimensionality is Singular Value Decomposition (SVD).

Compute the $U$, $D$, and $V^T$ elements from performing SVD on the matrix $A$ below. Do not round any values.
- Assign the $U$ matrix to `ans11a`
- Assign the $D$ matrix to `ans11b`
- Assign the $V^T$ matrix to `ans11c`
- Using the 2 largest singular values from the diagonal ($D$), assign the reduced matrix (shape: 5x2) to the variable `ans11d`. This matrix can be computed by taking the dot product of $U$ and $D$ after $D$ has been cast into a diagonal matrix with the same number of rows and columns as A.

Hints:
- You may use `svd` function from `scipy.linalg` package
- The array, $D$, returned from the bullet above can be cast to the correct shape by running:
```
diagonal = np.empty((A.shape[0], A.shape[1]))
diagonal[:A.shape[0], :A.shape[0]] = diag(D)
```

In [34]:
A = np.array([
    [8, 7, 3, 1, 1, 9, 9, 5, 7, 6],
    [9, 8, 1, 4, 4, 8, 9, 5, 9, 3],
    [0, 4, 1, 8, 4, 9, 3, 4, 2, 7]
])

In [35]:
### GRADED

### YOUR SOLUTION HERE
ans11a = None
ans11b = None
ans11c = None
ans11d = None

### BEGIN SOLUTION
# perform svd with scipy.linalg.svd()
U, D, VT = svd(A)

# make 'D' a diagonal matrix
diagonal = np.empty((A.shape[0], A.shape[1]))
diagonal[:A.shape[0], :A.shape[0]] = np.diag(D)

# assign values
ans11a = U
ans11b = diagonal
ans11c = VT
ans11d = U.dot(diagonal)
### END SOLUTION

### For verifying answer:
print('-'*40)
print("Original matrix")
print(A)
print('-'*40)
print("The 'U' matrix after decomposition")
print(ans11a)
print('-'*40)
print("The 'D' matrix after decomposition")
print(ans11b)
print('-'*40)
print("The 'VT' matrix after decomposition")
print(ans11c)
print('-'*40)
print("The reduced dimensionality matrix (U dot product with diagonal)")
print(ans11d)

----------------------------------------
Original matrix
[[8 7 3 1 1 9 9 5 7 6]
 [9 8 1 4 4 8 9 5 9 3]
 [0 4 1 8 4 9 3 4 2 7]]
----------------------------------------
The 'U' matrix after decomposition
[[-0.62462913 -0.28752816  0.72606198]
 [-0.65832617 -0.30622969 -0.68762637]
 [-0.42005368  0.90749707 -0.00199267]]
----------------------------------------
The 'D' matrix after decomposition
[[31.11470867  0.          0.          0.25089832  8.30082998 -5.22211538
   0.10752785 -0.20565235 -8.22750838  0.03584262]
 [ 0.         10.17000204  0.          0.45830134 -0.10009038  0.32258355
   0.15366475  0.42722896  0.32258355 -0.06795799]
 [ 0.          0.          4.29487636  0.12688627  0.25089832  0.11274778
  -0.22002659  0.2150557  -0.05886937  0.57349632]]
----------------------------------------
The 'VT' matrix after decomposition
[[-0.35102269 -0.36379026 -0.09488333 -0.21270851 -0.15870785 -0.47144117
  -0.41159822 -0.26016606 -0.35794797 -0.27842552]
 [-0.49717714 -0.08186295

In [36]:
### BEGIN HIDDEN TESTS (10)
A_ = np.array([
    [8, 7, 3, 1, 1, 9, 9, 5, 7, 6],
    [9, 8, 1, 4, 4, 8, 9, 5, 9, 3],
    [0, 4, 1, 8, 4, 9, 3, 4, 2, 7]
])

U_, D_, VT_ = svd(A_)

diagonal_ = np.empty((A_.shape[0], A_.shape[1]))
diagonal_[:A_.shape[0], :A_.shape[0]] = np.diag(D_)

ans11a_ = U_
ans11b_ = diagonal_
ans11c_ = VT_
ans11d_ = U_.dot(diagonal_)

student = [ans11a, ans11b, ans11c, ans11d]
solution = [ans11a_, ans11b_, ans11c_, ans11d_]

for stu, sol in zip(student, solution):
    np.testing.assert_equal(stu, sol, "The matrix is not correct", verbose = False)
print("That's correct!")
### END HIDDEN TESTS

That's correct!


[Back to top](#Index:) 

## %subtitle_4%
One ubiquitous use of linear algebra is to the solve least squares regression problem. This approach involves using a system of linear equations to solve for the beta coefficents directly. For an intuitive understanding, the below walks you through arriving at the normal equation from a regression model purely from an algebraic standpoint. See [this post](https://math.stackexchange.com/questions/644834/least-squares-in-a-matrix-form) for a complete derivation of the normal equations.

Let's review the least squares regression line equation: 

<br>

$\hat y = \hat \beta_0 + \hat \beta_1x_1 + ... + \hat \beta_nx_n$

<br>

...which can be rewritten in matrix notation as:

<br>

$\hat y = X \hat \beta$

<br>

where:

- $\hat \beta$ are the least squares beta coeffients
- $X$ is a matrix of data
- $\hat y$ is a vector of least squares predictions
<br>


where $\hat y$ are our model's predictions.

<br>

Using this notation, we can apply the following linear algebra operations


**Start**

<br>

$\hat y = X\hat\beta$

<br>

**Multiply both sides by $X^T$**

<br>

$X^T\hat y = X^TX\hat\beta$


<br>

**Multiply both sides by $(X^TX)^{-1}$**

<br>

$(X^TX)^{-1}X^T\hat y = (X^TX)^{-1}(X^TX)\hat\beta$

<br>

**Since $(X^TX)^{-1}(X^TX)$ is equal to the identity matrix, the right side of the equation simplifies down to:**

<br>

$(X^TX)^{-1}X^T\hat y = \hat\beta$

<br>

This final equation is known as the Normal Equation. Let's demonstrate this calculation with sample data


Consider the sample dataset below.


| Obs | x1 | x2 | y |
|:-:|:-:|:-:|:-:|
|**1**|   0.96    |    0.10   |  61.5  |
|**2**|   1.45    |    1.05   | 113.4  |
|**3**|   0.16    |    0.51   |  21.9  |
|**4**|  -0.31    |   -0.99   | -42.0  |
|**5**|   1.32    |   -0.77   |  64.5  |
|**6**|  -1.07    |   -1.43   | -98.6  |
|**7**|   0.56    |    0.29   |  41.4  |
|**8**|  -1.62    |    0.21   | -94.9  |
|**9**|   0.67    |    1.88   |  84.7  |
|**10**|  -0.48    |    0.85   | -10.2  |


[Back to top](#Index:) 

### Question 12

*15 Points*

For this question, you will be asked to use the Normal Equation above to solve for the vector `b`.

Using the array `A` and vector `b`, compute the parameter vector `x` and assign the result to the variable `ans12`. Round all values to 3 decimal places.

In [37]:
# independent variables
A = np.array([
    [ 0.96,  0.1 ],
    [ 1.45,  1.06],
    [ 0.17,  0.52],
    [-0.32, -0.99],
    [ 1.33, -0.77],
    [-1.07, -1.44],
    [ 0.56,  0.3 ],
    [-1.63,  0.22],
    [ 0.68,  1.89],
    [-0.48,  0.85]
])

# dependent (target) variable
y = np.array([
    61.5,
    113.4,
    21.9,
    -42. ,
    64.2,
    -98.6,
    41.4,
    -95. ,  
    84.7, 
    -10.2
])

In [38]:
### GRADED

### YOUR SOLUTION HERE
ans12 = None

### BEGIN SOLUTION
# use the normal equation shown above
ans12 = np.round(np.linalg.inv(A.T @ A) @ A.T @ y, 3) # --> array([61.483, 22.724])
### END SOLUTION

In [39]:
### BEGIN HIDDEN TESTS (15)
ans12_ = np.linalg.inv(A.T @ A) @ A.T @ y

np.testing.assert_almost_equal(ans12_, ans12, decimal =2, verbose = False)
print("That's correct!")
### END HIDDEN TESTS

That's correct!
